Upgrade to LSP 6.0.0 and Monaco 0.19.3

RomanNikitenko opened the pull request:
#### What it does
Upgrade to:
- `LSP`: `6.0.0` 
- `theia/monaco-editor-core`:  `0.19.3`
- `vscode-ws-jsonrpc`: `0.2.0`
- [x] depends on https://github.com/TypeFox/monaco-languageclient/pull/199
- [x] depends on https://github.com/TypeFox/monaco-languageclient/pull/191
- [x] support Codicons in the quick pick items https://github.com/eclipse-theia/theia/issues/5412
- [x] upgrade to the latest Monaco https://github.com/eclipse-theia/theia/issues/6900

#### TODO 
- [x] align signatures for `monaco/index.d.ts` https://github.com/eclipse-theia/theia/pull/7149#discussion_r379299301
- [x] add `Peek` submenu and the corresponding actions https://github.com/eclipse-theia/theia/pull/7149#issuecomment-593831885
- [x] provide proper column at `go to definition` https://github.com/eclipse-theia/theia/pull/7149#issuecomment-594411526
   - Provided [fix](https://github.com/eclipse-theia/theia/pull/7149/commits/c91b2130eca1f3e782eba1629e1d548cc71918c4)
   - applied tests, commented https://github.com/eclipse-theia/theia/pull/7265#pullrequestreview-369665248
- [x] remove `referencesController` and test `peek` widget https://github.com/eclipse-theia/theia/pull/7149#pullrequestreview-368634327
   - tested: https://github.com/eclipse-theia/theia/pull/7149#discussion_r388258781, need more investigation.
- [x] remove checks like: https://github.com/eclipse-theia/theia/pull/7149#pullrequestreview-368622518
   - removed, commented https://github.com/eclipse-theia/theia/pull/7149#discussion_r388280595
- [ ] revert the hotfix for IME support https://github.com/eclipse-theia/theia/pull/6673 and check whether IME support is fixed in newer Monaco
  - reverted the hotfix, but it's required more investigation why it doesn't work for `Theia`: https://github.com/eclipse-theia/theia/pull/6673#issuecomment-599999501  
- [x] check https://github.com/eclipse-theia/theia/issues/6390#issuecomment-542127645
  - tested: https://github.com/eclipse-theia/theia/issues/6390#issuecomment-595233432
 - [x]  `Peek Declaration` problem https://github.com/eclipse-theia/theia/pull/7149#issuecomment-599996086
   - investigated: https://github.com/eclipse-theia/theia/pull/7149#issuecomment-600843746
   - created the separate issue https://github.com/eclipse-theia/theia/issues/7368 as it's relevant for `master` branch as well
- [x] Fix `quick fixes` menu https://github.com/eclipse-theia/theia/pull/7149#issuecomment-600056386
   - provided fix: https://github.com/eclipse-theia/theia/pull/7149#issuecomment-600315543 
- [x] Go to Symbol commands have different title and keybindings to VS Code https://github.com/eclipse-theia/theia/pull/7149#issuecomment-601066986
   - renamed `Open Workspace Symbol...` => `Go to Symbol in Workspace...`: https://github.com/eclipse-theia/theia/pull/7149/commits/445104f1ffb6c958a636850e417dd872995318bf
   - commented https://github.com/eclipse-theia/theia/pull/7149#issuecomment-602059511
- [x] Selection menu in the main bar is not aligned with VS Code anymore https://github.com/eclipse-theia/theia/pull/7149#issuecomment-601104312
   - fixed https://github.com/eclipse-theia/theia/pull/7149#issuecomment-602193180
- [x] check that editor configurations are aligned https://github.com/eclipse-theia/theia/pull/7149#issuecomment-601114071 
   - investigated https://github.com/eclipse-theia/theia/pull/7149#issuecomment-602026175
   - updated editor configurations https://github.com/eclipse-theia/theia/pull/7149#issuecomment-603677700
- [x] test tasks integration for VS Code extensions
   - elaihau (thanks!) tested task palette https://github.com/eclipse-theia/theia/pull/7149#issuecomment-602258067
   - steps for testing https://github.com/eclipse-theia/theia/pull/7149#issuecomment-604067026
- [x] Workspace icons are missing
   - fixed by https://github.com/eclipse-theia/theia/pull/7149/commits/050166b16da2cfa35c9bdfb38206cd5d602363a6
- [x] check `CallHierarchy` https://github.com/eclipse-theia/theia/pull/7149#issuecomment-602022641
   - latest typescript VS Code extension is required https://github.com/eclipse-theia/theia/pull/7149#issuecomment-602463783
   - created issue https://github.com/eclipse-theia/theia/issues/7395
- [x] Apply codicons for `Configure Task` action 
   - applied https://github.com/eclipse-theia/theia/pull/7149#issuecomment-602084033
- [x] Test  `codicons` for command palette
   - found some issues https://github.com/eclipse-theia/theia/pull/7149#issuecomment-602688748
   - fix styles https://github.com/eclipse-theia/theia/pull/7149#issuecomment-602834354
   - created the issue for VS Code https://github.com/microsoft/vscode/issues/93303
   - created the issue for Theia https://github.com/eclipse-theia/theia/issues/7429


 

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

- [x] test language features with VS Code extension:
  - [x] test that features are available in the editor context menu and in quick command palette (automated)
  - [x] a declaration is revealed when following the reference (automated)
  - [x] focus belongs to a proper editor after following a reference (for internal and outside references, from editor to editor preview and view versa) (automated)
  - [x] reference peek is still opened after following a reference (for internal and outside references, from editor to editor preview and view versa) (automated)
  - [x] esc closes reference peek widget (automated)
  - [x] completion items are properly resolved and inserted (automated)
  - [x] focus belongs to a proper editor on peek definition (for internal and outside references, from editor to editor preview and view versa) (automated)
  - [x] check that rename work inside the file and across files (automated)
  - [x] check that arguments are proposed by the signature help (automated)
  - [x] check that hover is present with properly rendered docs (automated)
  - [x] check that symbol occurrences is highlighted when a declaration or reference is selected (automated)
  - [x] check that go to (type|implementation) definition works properly, i.e. definition is revealed (automated)
  - [x] check that code lens work properly (automated)
      - [x] run/debug code lenses: https://github.com/kasecato/vscode-javadebug-sample. One has to install vscode java, debug and test extensions.
  - [x] check that code actions work properly (quick fixes) (automated)
  - [x] check that document and selection formatting work properly (automated)
  - [x] check color provider with css colors
  - [x] check link provider with links in html
  - [x] check that folding ranges are properly provided for js, css and html
  - [x] check go to symbol command
  - [x] check open workspace symbol command
- [x] test quick palette:
  - [x] proper keys are displayed for the quick command palette, i.e. they are displayed and a corresponding Monaco command can be triggered by such keybinding
  - [x] use sample vscode extension
  - [x] use fuzzy search
- [x] test vscode tree view (that context keys work properly)
- [x] test plugin editor decorations, for example with power mode extension
- [x] test debug console input editor
- [x] test debug breakpoint editor
- [x] test that preferences are propagated as config key values: change editor preferences and check whether they are reflected
- [x] test language auto detection
- [x] test `edit` menu with editor
- [x] test `selection` menu with editor
- [x] test editor context menu
- [x] smoke test some VS Code extensions like go, python, just use editor for a bit, check backend and frontend logs for errors
- [x] test manually task related functionality https://github.com/eclipse-theia/theia/pull/7149#issuecomment-604067026

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: Roman Nikitenko <rnikiten@redhat.com>

akosyakov commented on packages/monaco/src/typings/monaco/index.d.ts:
```diff
@@ -15,7 +15,7 @@
 
 /* eslint-disable @typescript-eslint/no-explicit-any */
-/// <reference types='@typefox/monaco-editor-core/monaco'/>
+/// <reference types='@theia/monaco-editor-core/monaco'/>
```
@RomanNikitenko now, the tedious part is coming 😬 

Usually we go through this file and check that signatures exist and matching in https://github.com/theia-ide/vscode/tree/standalone/0.19.x If not they should be aligned

RomanNikitenko replied:
@akosyakov 
Should we follow **strictly**  the corresponding signatures of vs code?
Or the answer is: it depends?

For example, `IBulkEditService` [in current](https://github.com/eclipse-theia/theia/blob/1eaecd5cbf93468a74988ddcd347402fcbf14cbe/packages/monaco/src/typings/monaco/index.d.ts#L32) `index.d.ts` differs  from [0.17 version](https://github.com/theia-ide/vscode/blob/63d87164d0bc8c6206d9339c195289c93665028e/src/vs/editor/browser/services/bulkEditService.ts#L23) of `IBulkEditService` 

0.17 version `IBulkEditService` contains:
- field `_serviceBrand`
- `apply` method has `options?: IBulkEditOptions` as optional parameter.

So, should we align it now for 0.19.3? 
Should we follow **strictly**  the corresponding signatures of vs code in other cases?

akosyakov replied:
We should expose only what we use, if it is not used, then it should not be exposed.

akosyakov commented:
@RomanNikitenko Could you reference relevant issue and fill int `How to test` with boxes from previous migration please?

RomanNikitenko commented:
@akosyakov 
I updated the PR.

RomanNikitenko commented:
I found that `Peek` actions are coming as items of `Peek` submenu in `vs code`, so I'm going to align it for `theia`.

akosyakov commented:
@RomanNikitenko ok, I will start testing it a bit this week.

akosyakov commented:
btw thank you for such diligent work on updating monaco typings 👍

akosyakov commented:
@RomanNikitenko I've started adding tests in https://github.com/eclipse-theia/theia/pull/7265 against this PR. I will stabilise and add more tests next days. They should cover `How to test` section of this PR eventually.

RomanNikitenko commented:
> @RomanNikitenko I've started adding tests in #7265 against this PR. I will stabilise and add more tests next days. They should cover `How to test` section of this PR eventually.

I think it will really help to test the current changes and will speed up the process of testing of next Monaco upgrade!

Thank you, Anton for doing this. l really appreciate it! 👍

akosyakov commented:
I've noticed that go to definition does not jump to proper column anymore:
- put a cursor in theia/examples/browser/src-gen/backend/server.js at Container word
- use go to definition
- on master a new editor is revealed and cursor is put at Container word with this PR it ends up at the beginning of line

akosyakov commented on packages/monaco/src/browser/monaco-editor-provider.ts:
```diff
@@ -176,6 +182,11 @@ export class MonacoEditorProvider {
     protected injectKeybindingResolver(editor: MonacoEditor): void {
         const keybindingService = editor.getControl()._standaloneKeybindingService;
+        if (!keybindingService) {
+            console.warn('Cannot inject Keybinding Resolver because KeybindingService is not recognized');
```
Could it happen in the normal application that it is undefined? If not then we should fail, i think it is fine without a guard it should fail with undefined reference exception and it has to be fixed. Please remove such checks if there are more.

RomanNikitenko replied:
I added such checks because `_standaloneKeybindingService` [became nullable](https://github.com/microsoft/vscode/commit/31c73988b9fe71ea10ebf881d0a178f3d8e14743) and VS Code [has similar checks](https://github.com/microsoft/vscode/blob/31c73988b9fe71ea10ebf881d0a178f3d8e14743/src/vs/editor/standalone/browser/standaloneCodeEditor.ts#L190).

I removed this checks for `theia`: https://github.com/eclipse-theia/theia/pull/7149/commits/02ae87b8f87849e4b787e2d78f461ea57c163848

akosyakov commented on packages/monaco/src/browser/monaco-editor-provider.ts:
```diff
@@ -353,8 +364,10 @@ export class MonacoEditorProvider {
     protected installReferencesController(editor: MonacoEditor): void {
```
Maybe we don't need it anymore, It seems that this code was fixed in VS Code: https://github.com/spahnke/vscode/commit/b9f4d37650e21167eb04747e8b380598d6f9f5dd

Let's try to remove it and test that peek widget is not closed while changing editors.

RomanNikitenko replied:
I tried to remove it and found that it doesn't work correctly. It works when I open a first editor, but the dialog is closed when I try to open a next editor.

![peek_reference_bug](https://user-images.githubusercontent.com/5676062/75981080-66e81500-5eec-11ea-9dc3-69b80c876784.gif)

So it's still in the PR `TODO` list  as unresolved. I will investigate why it works for `vs code`, but doesn't work for `theia`.

RomanNikitenko replied:
Tested and debugged against https://github.com/theia-ide/vscode/tree/standalone/0.19.x.
`referencesController` [works well](https://github.com/theia-ide/vscode/blob/standalone/0.19.x/src/vs/editor/contrib/gotoSymbol/peek/referencesController.ts#L231) tor this case.
Investigating `theia` side...

RomanNikitenko replied:
For case when `peek` widget is open VS Code doesn't create `codeEditorWidget` every time an editor opens. As result [the condition](https://github.com/theia-ide/vscode/blob/standalone/0.19.x/src/vs/editor/contrib/gotoSymbol/peek/referencesController.ts#L229) is met and `peek` widget is open.
Theia creates  `standaloneCodeEditor` instance every time, so the condition [is NOT met](https://github.com/theia-ide/vscode/blob/standalone/0.19.x/src/vs/editor/contrib/gotoSymbol/peek/referencesController.ts#L234), so `peek` widget on the next iteration [is closed](https://github.com/theia-ide/vscode/blob/standalone/0.19.x/src/vs/editor/contrib/gotoSymbol/peek/referencesController.ts#L225).
I think that's why we have this problem for `Theia` and don't have for `VS Code`

akosyakov replied:
ok, let's keep our fix

akosyakov commented:
Regarding to https://github.com/eclipse-theia/theia/pull/7149#issuecomment-594411526

@RomanNikitenko there is a bug in the plugin system which breaks navigation, here https://github.com/eclipse-theia/theia/blob/a0e57d0b4f358e526d04ca85429575e19e688178/packages/plugin-ext/src/common/plugin-api-rpc-model.ts#L308-L313 there is `selectionRange`, but there is no such property in vscode.d.ts and monaco, instead it should be `targetSelectionRange` and `originSelectionRange`

I will add an integration test for it.

RomanNikitenko commented:
@akosyakov 
> there are selectionRange and origin, but there are no such properties monaco.d.ts, instead it should be targetSelectionRange and originSelectionRange

the corresponding interface [was renamed](https://github.com/microsoft/vscode/commit/14ce5183db4d016383a779b436c30b434c1f86c0#diff-2ad890a42cb36d54a33f89dfea7330f6R719) in vs code: `DefinitionLink` => `LocationLink`

Should we rename it as well in `plugin-api-rpc-model` and keep it close to `vs code`?
or
Do you prefer to keep an old name for such cases?

akosyakov commented:
@RomanNikitenko I would be fine with it, it is not public API, so should be fine to align. It is certain helps when things named the same way during investigations.

akosyakov commented:
@RomanNikitenko thank you for the update, i will continue adding test tomorrow, had to focus on the vsx registry integration today. Could you check other my comments in the meantime?

RomanNikitenko commented:
@akosyakov
- added `Peek` submenu
- added fix for `go to definition`. Checked your use case manually - looks like it's fixed now. Applied your test - it failed. Will take a look at this today more closely...   

>  Could you check other my comments in the meantime?

sure! thank you for your comments - I added `TODO` section to the PR description to not lose your comments and fix them.

I'm not sure I understand the use case for testing here: https://github.com/eclipse-theia/theia/pull/7149#discussion_r387553989. I played with it yesterday - tried to add and remove the corresponding code, but I didn' see the difference. Will do it today again...

akosyakov commented:
> I'm not sure I understand the use case for testing here: #7149 (comment). I played with it yesterday - tried to add and remove the corresponding code, but I didn' see the difference. Will do it today again...

@RomanNikitenko You can try to remove it on master and see what is the difference. I think it was related to peek reference dialog if you follow a reference from it then a new editor should be opened and it should have the peek reference opened as well, but old widget should not. Plus focus should be transferred properly, i.e. try to move up/down or type after navigation. On master it should break when you remove with new Monaco i hope it should b fine.

RomanNikitenko commented:
@akosyakov 
> You can try to remove it on master and see what is the difference.

Got it! Yes, now I can reproduce it, thank you!
I added comment https://github.com/eclipse-theia/theia/pull/7149#discussion_r388258781

akosyakov commented:
@RomanNikitenko Unfortunately I did not do enough progress with the Open VSX Registry integration and would prefer to focus today again on it. I hope I can invest more time in testing this PR on monday.

I hope you can have a look at removing IME workaround and ping relevant folks to test as well.

RomanNikitenko commented:
@akosyakov 
No problem, I think I have enough tasks for today and I'm going to be offline on Monday.
So have a nice day and see you on Tue!
:-)

akosyakov commented:
@RomanNikitenko Could you rebase please?

akosyakov commented:
@RomanNikitenko I've noticed that VS Code does not show `Peek Declaration` always, but our `Peek` submenu show it always. Do we miss some `when` closure for this menu item?

RomanNikitenko commented:
@akosyakov 
yes, I noticed that as well.
Will investigate it today...

akosyakov commented:
@RomanNikitenko I cannot trigger quick fixes. It always opens the context menu instead. It does no work from the light bubble hint as well as from the error hover.

RomanNikitenko commented:
> I cannot trigger quick fixes. It always opens the context menu instead. It does no work from the light bubble hint as well as from the error hover.

thank you for the use case!
Looks like [the condition](https://github.com/eclipse-theia/theia/blob/9c4e4d9ce9f10c70ac53f77b321a21d2b5576168/packages/monaco/src/browser/monaco-context-menu.ts#L35) is not met after upgrade `Monaco` version.
I provided the fix, please take a look when you have a chance.

502647092 commented:
![github](https://user-images.githubusercontent.com/10203793/76915015-e1eef980-68f6-11ea-9991-139ddc0f76e4.gif)
Chinese IME test result

mouse484 commented:
@502647092 
Try with this commit -> 32b26b27a02a0c21e7c5a17ad1bed5b63ed9c975 

Maybe works

akosyakov commented:
> I provided the fix, please take a look when you have a chance.

@RomanNikitenko It seems to work. I will automate checking it today.

akosyakov commented:
@502647092 @mouse484 Sounds like we should keep a workaround? I wonder why it works in the vanilla Monaco but not for us 🤔

RomanNikitenko commented:
@akosyakov 
> I wonder why it works in the vanilla Monaco but not for us

I'm investigating `Peek Declaration` issue at the moment and then will take a look at IME support. At least to know what's the cause of the problem

akosyakov commented on packages/monaco/src/browser/monaco-context-menu.ts:
```diff
@@ -30,12 +30,10 @@ export class MonacoContextMenuService implements IContextMenuService {
-        // Unfortunately checking the existence of certain properties seems to be the best way to tell, what kind of context menu is requested.
-        if (delegate.hasOwnProperty('getKeyBinding')) {
+        const actions = delegate.getActions();
+        if (actions.length > 0 && actions[0] instanceof monaco.actions.MenuItemAction) {
```
@RomanNikitenko Could you add a comment why condition is like that? How actions look like for quick fixes?

RomanNikitenko replied:
In case of editor context menu actions come as `MenuItemAction`
In case of `Quick Fix` actions come as `CodeActionAction`

RomanNikitenko replied:
Is this way good enough? If so, I'll add the corresponding comment.

akosyakov replied:
yes, it is good

akosyakov commented on packages/monaco/src/typings/monaco/index.d.ts:
```diff
@@ -364,6 +364,25 @@ declare module monaco.actions {
+
+    // https://github.com/theia-ide/vscode/blob/standalone/0.19.x/src/vs/platform/actions/common/actions.ts#L246
+    export class MenuItemAction {
+        readonly item: ICommandAction;
```
Do we need to copy all properties?  I don't see them to be used.

RomanNikitenko replied:
I removed properties.
I use this class only to check `actions[0] instanceof monaco.actions.MenuItemAction`

akosyakov replied:
probably constructor is not necessary as well?

RomanNikitenko replied:
so, just empty class?
:-)

RomanNikitenko commented:
@akosyakov 
> @RomanNikitenko I've noticed that VS Code does not show Peek Declaration always, but our Peek submenu show it always. Do we miss some when closure for this menu item? 
 
I investigated the case and I'd say it's not `Peek Declaration` issue, but more global issue and it’s relevant for the `master` branch as well.
 
VS Code recreates editor context menu on each call and adds the corresponding actions based on `contextMatchesRules(item.when)`.
Theia creates editor context menu only once. 
 
As result we have the same context menu for different editors: https://youtu.be/kmZeuxAfe_I
 
I think we could recreate a context menu based on given actions (from `IContextMenuDelegate`)  on each call.
I can create a separate issue for that and try to implement it if you think it’s true way and we don’t have more urgent issues at the moment related to upgrade Monaco.

akosyakov commented:
> I can create a separate issue for that and try to implement it if you think it’s true way and we don’t have more urgent issues at the moment related to upgrade Monaco.

@RomanNikitenko it sounds fine, let's have just an issue for it now.

akosyakov commented:
@RomanNikitenko Could you rebase again please?

akosyakov commented:
`Go to Symbol` commands have different title and keybindings to VS Code.

VS Code:
<img width="633" alt="Screenshot 2020-03-19 at 10 03 37" src="https://user-images.githubusercontent.com/3082655/77049765-fac7df80-69c8-11ea-97c1-a9a07b6c3b18.png">
Theia:
<img width="612" alt="Screenshot 2020-03-19 at 10 03 45" src="https://user-images.githubusercontent.com/3082655/77049770-fc91a300-69c8-11ea-8504-8e815b3f06e0.png">

For workspace icons are missing:
<img width="628" alt="Screenshot 2020-03-19 at 10 03 58" src="https://user-images.githubusercontent.com/3082655/77049773-fd2a3980-69c8-11ea-9c9e-d77a363d5d68.png">

Also order is different to VS Code for worspace symbols, but I think it should be fixed separately. It seems to be the same on master.

I will add a regression test for missing icons.

akosyakov commented:
`Selection` menu in the main bar is not aligned with VS Code anymore.

VS Code:
<img width="316" alt="Screenshot 2020-03-19 at 11 27 15" src="https://user-images.githubusercontent.com/3082655/77057715-a9bde880-69d4-11ea-9a62-7e7dfdc06786.png">

Theia:
<img width="381" alt="Screenshot 2020-03-19 at 11 27 20" src="https://user-images.githubusercontent.com/3082655/77057724-ad516f80-69d4-11ea-8508-f7aa2f6be50e.png">

- `Expand Selection` missing
- `Duplicate Selection` missing
- occurrences related items are named differently missing

akosyakov commented:
I went through rest of tests manually. I think we can automate them next time. I will try to add tests only for regressions (if it is possible, i.e comparing to VS Code can be done only manually).

@RomanNikitenko I think we should start wrapping up this PR. It generally looks good. The worst issue is missing icons for workspace go to symbols. Others are nice to fix but can be resolved with follow-ups. It could be also that the difference to VS Code is because Monaco 0.19 is not latest version. Regarding IME support workaround let's keep it as is and investigate separately.

akosyakov commented:
I will do a release of monaco-languageclient today that we can upgrade to stable version.

akosyakov commented on packages/monaco/src/browser/monaco-menu.ts:
```diff
@@ -66,6 +66,8 @@ export namespace MonacoMenus {
     ];
 }
 
+export const PEEK_CONTEXT_SUBMENU: MenuPath = [...EDITOR_CONTEXT_MENU, 'navigation', 'peek_submenu'];
```
It should go under MonacoMenus

RomanNikitenko replied:
done

akosyakov commented:
Ah I remember one more thing, we should check that editor configurations are aligned here: https://github.com/eclipse-theia/theia/blob/master/packages/editor/src/browser/editor-preferences.ts

akosyakov commented:
@RomanNikitenko monaco-languageclient@0.12.0 is released

RomanNikitenko commented:
@akosyakov 
> Ah I remember one more thing, we should check that editor configurations are aligned here: https://github.com/eclipse-theia/theia/blob/master/packages/editor/src/browser/editor-preferences.ts

editor configurations was reorganized in VS Code: https://github.com/theia-ide/vscode/commit/b71b8a48ded639b56ca4027e41e863b904053659
working on adaption for Theia...

akosyakov commented:
@RomanNikitenko Could you rebase again please? If you can rebasing each day would help to make sure that everything works against master, thank you!

RomanNikitenko commented:
@akosyakov 
is it useful for review to keep history of commits?
Can I squash my 14 commits in 1?

akosyakov commented:
> is it useful for review to keep history of commits? Can I squash my 14 commits in 1?

@RomanNikitenko While reviewing, for me it is fine any way :) At the end it will be good to fixup commits which fixes issues introduced by the PR itself.

akosyakov commented:
@RomanNikitenko I did not test tasks integration for VS Code extensions. Could you have a look? I think we rely on some Monaco apis to show actions in the quick task palette. Maybe @elaihau can smoke test as well.

RomanNikitenko commented:
@akosyakov 
> I did not test tasks integration for VS Code extensions. Could you have a look?

will do it, added to my `todo` list in the PR description

elaihau commented:
> @RomanNikitenko I did not test tasks integration for VS Code extensions. Could you have a look? I think we rely on some Monaco apis to show actions in the quick task palette. Maybe @elaihau can smoke test as well.

@akosyakov sure i can do some smoke testing for the task palette in the next few days

RomanNikitenko commented:
@akosyakov 
Does `CallHierarchy` work in `Theia`?

I tried for `master` branch as well - it always disabled...

RomanNikitenko commented:
@akosyakov 
About https://github.com/eclipse-theia/theia/pull/7149#issuecomment-601425304

`editorConfiguration` in VS Code contains [limited number of properties](https://github.com/theia-ide/vscode/blob/standalone/0.19.x/src/vs/editor/common/config/commonEditorConfig.ts#L440) now.
Other properties [are added](https://github.com/theia-ide/vscode/blob/standalone/0.19.x/src/vs/editor/common/config/commonEditorConfig.ts#L526) from `editorOptionsRegistry` which come from [EditorOptions](https://github.com/theia-ide/vscode/blob/standalone/0.19.x/src/vs/editor/common/config/editorOptions.ts#L3132)

We could read them in `editor-preferences.ts` like:
```
const options: { [key: string]: any; } = monaco.editor.EditorOptions;
for (const key in options) {
    const editorOption = options[key];
    const schema = editorOption.schema;
    if (typeof schema === 'undefined') {
        continue;
    }

    if (typeof schema.type !== 'undefined' || typeof schema.anyOf !== 'undefined') {
        codeEditorPreferenceProperties.properties[`editor.${editorOption.name}`] = schema;
    } else {
        for (let key in schema) {
            if (Object.hasOwnProperty.call(schema, key)) {
                codeEditorPreferenceProperties.properties[key] = schema[key];
            }
        }
    }
}
```


But I guess this is a bad idea, because this module shouldn’t depend on `monaco`.

So should we instead copy [EditorOptions](https://github.com/theia-ide/vscode/blob/standalone/0.19.x/src/vs/editor/common/config/editorOptions.ts#L3132) and clean up them(remove `nls.localize...` and so on)?
Maybe you see another solution.

thanks in advance!

RomanNikitenko commented:
@akosyakov 
> For workspace icons are missing:

Should be fixed by https://github.com/eclipse-theia/theia/pull/7149/commits/050166b16da2cfa35c9bdfb38206cd5d602363a6

RomanNikitenko commented:
@akosyakov 
> Go to Symbol commands have different title and keybindings to VS Code.

I renamed `Open Workspace Symbol...` => `Go to Symbol in Workspace...`: https://github.com/eclipse-theia/theia/pull/7149/commits/445104f1ffb6c958a636850e417dd872995318bf

I guess `Theia` uses `Ctrl+O` because there is no way to override `Ctrl+T` in Chrome.

About `Go to Symbol in File`:
I found that `Monaco Editor Playground` displays it as `Go to Symbol...` as well:

![Go_to_symbol](https://user-images.githubusercontent.com/5676062/77229680-e4bc4980-6b97-11ea-94fd-7540a4b4c1e0.gif)

So, maybe it's expected behavior.
Please let me know if I should investigate how we can customize it for `Theia` and align with VS Code.

RomanNikitenko commented:
@akosyakov 
I removed `svg` icons for `Configure Task` action and applied `codicon`: https://github.com/eclipse-theia/theia/pull/7149/commits/afc03a29d6e9af4a437cbdbd7708346559f69215
Also hover was added for the action.

![Dark_configure](https://user-images.githubusercontent.com/5676062/77233810-b8fa8d00-6bb2-11ea-83f9-3455a3ff330d.gif)

RomanNikitenko commented:
@akosyakov 
About https://github.com/eclipse-theia/theia/pull/7149#issuecomment-601104312
> Selection menu in the main bar is not aligned with VS Code anymore.

I think VS Code uses:
-  [label](https://github.com/theia-ide/vscode/blob/standalone/0.19.x/src/vs/editor/contrib/multicursor/multicursor.ts#L663) for command palette, so `Add Selection To Next Find Match` 
but 
- [title](https://github.com/theia-ide/vscode/blob/standalone/0.19.x/src/vs/editor/contrib/multicursor/multicursor.ts#L674) for menu, so `Add Next Occurrence`

I tried to apply the same behavior for `Theia`:
![selection](https://user-images.githubusercontent.com/5676062/77249389-6f10b600-6c49-11ea-8d29-0730f527fd13.gif)

Please take a look at: https://github.com/eclipse-theia/theia/pull/7149/commits/0990f10ab2c02686039720d9fa4fd37306d94fcf
Pay attention that a titles can come with mnemonic, so I have to clean up them.  

Is it a good way from your point of view?

elaihau commented:
I played with this change in Chrome, Ubuntu 18.04. The task palette works properly. And i love how the new "Configure" icon looks like

akosyakov commented:
> So should we instead:
> - copy EditorOptions to our code and clean up them(remove nls.localize... and so on)?
> - or extract required properties manually from EditorOptions, update codeEditorPreferenceProperties > and keep our current format
> - or maybe you see another solution.

@RomanNikitenko the current format was achieved by copying from VS Code and removing `nls.localize`. We should do something similar again.

Long term we should think whether there is point into having 2 extensions `@theia/editor` and `@theia/monaco`. An original idea was to protect against Monaco internal changes and provide independent API but it causes maintenance burden and in practice many extensions still grab `@theia/monaco` directly. Maybe there is a way to have own extension that we could make use of Monaco code directly and still have buffed layer on top of Monaco internal APIs. It could be one of aspects to consider for https://github.com/eclipse-theia/theia/issues/7261

akosyakov commented:
> Does CallHierarchy work in Theia?

@RomanNikitenko I think we should migrate to latest typescript VS Code extension to support it. It is not related to your PR. You can file an issue though.

akosyakov commented:
I've tested again. It looks like aligning editor preferences is left. Let's target to merge it after Theia v1.0 release, so on Friday probably.

RomanNikitenko commented:
Within testing `codicons` for command palette I found the following issues:
- css alignment for some icons
- an icon is not rendered if this one contains word which you are typing. For example, you are trying to find `Close Workspace` item which contains `$(chrome-close)` for rendering. You are typing `close` - the icon is not rendered.  

![codicons_issue](https://user-images.githubusercontent.com/5676062/77335559-ff74f680-6d2e-11ea-9670-c33f58837354.gif)

RomanNikitenko commented:
Added style for `quick-open-entry` to fix codicons alignment https://github.com/eclipse-theia/theia/pull/7149/commits/969ce9000ab02e303b2c62bbeddfeb06fa8d29fc

![style_codeIcons](https://user-images.githubusercontent.com/5676062/77359630-8b4d4980-6d54-11ea-9561-af327a1bbd35.gif)

akosyakov commented:
> I can see the same behavior for Monaco Playground

@RomanNikitenko Please file an issue for VS Code.

akosyakov commented:
@RomanNikitenko I will continue today with stabilising typescript integration tests, let me know if you need me to have a look at something in this PR.

RomanNikitenko commented:
@akosyakov 
I'll try to take care of editor configurations alignment today

akosyakov commented:
@RomanNikitenko @elaihau if you did some manual testing which revealed issues please add them to `How to test` section for next time.

RomanNikitenko commented:
@akosyakov 
Properties `editor.gotoLocation.multiple` and `editor.suggest.filteredTypes` are coming as deprecated and without `default`. I had to add empty values for the `default` field to avoid compilation errors.

I guess we could remove them. 
On the other hand, it allows to display warning message for user if he/she tries to add them.
Also it informs user which property he/she should use instead.

![preference_deprec](https://user-images.githubusercontent.com/5676062/77510762-2b4fc380-6e78-11ea-83f3-6c95c2ae9de3.gif)

RomanNikitenko commented:
Regarding `codicons` rendering problems [described in the comment](https://github.com/eclipse-theia/theia/pull/7149#issuecomment-602688748) - 
looks like the problem on VS Code side, created the corresponding issue https://github.com/microsoft/vscode/issues/93303

akosyakov commented on packages/editor/src/browser/editor-preferences.ts:
```diff
@@ -59,223 +59,20 @@ export const EDITOR_MODEL_DEFAULTS = {
 // 3. Find -> Use Regular Expressions -> nls\.localize\(.*, '(.*)'\) -> '$1'
-// 4. Apply `quotemark` quick fixes
-// 5. Fix the rest manually
+// 4. Find -> Use Regular Expressions -> nls\.localize\(.*, '(.*)'\) -> '$1'
```
3 and 4 the same?

akosyakov commented:
@RomanNikitenko I see that there `migrateOptions` in VS Code: https://github.com/theia-ide/vscode/blob/d24b5f70c69b3e75cd10c6b5247a071265ccdd38/src/vs/editor/common/config/commonEditorConfig.ts#L146-L273

does it mean that some old settings files won't work anymore? If so we should apply migration as well.

akosyakov commented:
@RomanNikitenko I've tested new editor preferences and it is good. Is anything else left? ready for review?

RomanNikitenko commented:
@akosyakov 
> I see that there `migrateOptions` in VS Code

I'll check, thank you for pointing that out!

akosyakov commented:
> I'll check, thank you for pointing that out!

@RomanNikitenko you don't need to, I've checked, it is good. Everything is handled by Monaco already.

marcdumais-work commented:
> Please don't merge before 1.0 release! @marcdumais-work Could you give your approve after the release, thank you!

:+1:  Sounds good - let's merge after the release.

marcdumais-work commented:
@akosyakov normally we prefer to squash numerous commits in PRs. Would this one be a worthy exception to make or would you advise squashing?

RomanNikitenko commented:
I was trying to keep the history of commits on review step.
Sometimes it's useful for review and for case if we need to revert something or get back some changes again and so on.

Now I can squash it, no problem from my side.

akosyakov commented:
@RomanNikitenko as you wish, i think it is fine to keep commits which addressed separate problems on master, but commits which fix issues caused by the PR should be fixed up.

It depends on your motivation and how good you are with interactive rebase :)

RomanNikitenko commented:
About tasks for `How to test` section, to detect some `QuickOpenModel` related problems:

Go to Terminal -> Run Task:
-  task palette should display `detected`, `configured` and `recently used` tasks
- `codicon-gear` is displayed for a selected item
- hover `Configure Task` is displayed when select `codicon-gear`
- the corresponding `tasks.json` file is open at click on `codicon-gear` and the configuration of selected task is available for editing
- run tasks from different sections (`detected`, `configured`, `recently used`) - a terminal widget  with output should be open at running task

Additionally you can check:
- try to use other task related actions from `Terminal` menu
- try to create a new configuration in `tasks.json` file - task schema support should work
- `Problems` view should display warnings for a wrong configuration in `tasks.json` file

marcdumais-work commented:
Hi,

I tried this PR on Gitpod and while testing whether latest version of extension `vscode-clangd` would work now in Theia with LSP 6.0.0 support, I encountered this exception about `registerSelectionRangeProvider()`:

```
root ERROR [hosted-plugin: 4849] Promise rejection not handled in one second: TypeError: vscode_1.languages.registerSelectionRangeProvider is not a function , reason: TypeError: vscode_1.languages.registerSelectionRangeProvider is not a function
With stack trace: TypeError: vscode_1.languages.registerSelectionRangeProvider is not a function
    at SelectionRangeFeature.registerLanguageProvider (/home/gitpod/.theia/extensions/llvm-vs-code-extensions.vscode-clangd-0.0.18/extension/node_modules/vscode-languageclient/lib/selectionRange.js:45:35)
    at SelectionRangeFeature.register (/home/gitpod/.theia/extensions/llvm-vs-code-extensions.vscode-clangd-0.0.18/extension/node_modules/vscode-languageclient/lib/client.js:685:29)
    at SelectionRangeFeature.initialize (/home/gitpod/.theia/extensions/llvm-vs-code-extensions.vscode-clangd-0.0.18/extension/node_modules/vscode-languageclient/lib/selectionRange.js:30:14)
    at ClangdLanguageClient.initializeFeatures (/home/gitpod/.theia/extensions/llvm-vs-code-extensions.vscode-clangd-0.0.18/extension/node_modules/vscode-languageclient/lib/client.js:2385:21)
    at connection.initialize.then (/home/gitpod/.theia/extensions/llvm-vs-code-extensions.vscode-clangd-0.0.18/extension/node_modules/vscode-languageclient/lib/client.js:2081:18)
```
Is this expected?

marcdumais-work commented:
Looks this would be handled in https://github.com/eclipse-theia/theia/issues/6804 ?

akosyakov commented:
@marcdumais-work It is expected, we don't support such API in the plugin system yet: https://github.com/eclipse-theia/theia/issues/6623

RomanNikitenko commented:
@akosyakov 
btw: about `codicons` rendering issue https://github.com/microsoft/vscode/issues/93303#issuecomment-603937372

akosyakov commented:
@RomanNikitenko So we need a new monaco uprade :) ?

RomanNikitenko commented:
I think `codicons` rendering issue wasn't fixed, instead `codicons`  for VS Code editor actions and commands was disabled.

Please correct me if I'm wrong.

akosyakov commented:
@RomanNikitenko but only for command palette, it should still work generally for quick palette? It will break VS Code extensions otherwise.

RomanNikitenko commented:
@akosyakov 
> but only for command palette, it should still work generally for quick palette? 

yes, I think so, I meant we should investigate how to fix rendering issue, as it wasn't fixed.

akosyakov commented:
> yes, I think so, I meant we should investigate how to fix the rendering issue on our own, as it wasn't fixed.

Why should we? It does not work in VS Code and Monaco, why should it work in Theia?

RomanNikitenko commented:
@akosyakov 
For example, for VS Code:
![VsCode_git](https://user-images.githubusercontent.com/5676062/77641585-f3727a00-6f64-11ea-9836-3277383f6185.gif)

To have the same icon for Theia I go to `git-quick-open-service` and added `codicon` to the label
 
![Theia_git](https://user-images.githubusercontent.com/5676062/77641739-3a606f80-6f65-11ea-8bef-5929105a8ab4.gif)

It's bad example to type `plus` for this case, but it's just for testing goals.
But `$(link) Open Link` is better example. 

Could you advise some VS Code extension which use `quick-pick` items as good example to test it for Theia? I'll test and report the result here.

akosyakov commented:
> For example, for VS Code:

Are you sure that you are looking at VS Code with https://github.com/microsoft/vscode/commit/51e5259b11b23178e98879b7767306ad0ca188b0? Maybe it is already in VS Code insider.

> Could you advise some VS Code extension which use quick-pick items as good example to test it for Theia? I'll test and report the result here.

Unfortunately I could not find such among built-in and sample vscode extensions, but from [vscode.d.ts](https://github.com/microsoft/vscode/blob/85d2077f1924625bd2a89aa3ef5af26108c4c27d/src/vs/vscode.d.ts#L792-L816):

> A reference to a named icon. Currently, [File](#ThemeIcon.File), [Folder](#ThemeIcon.Folder), and [codicons](https://microsoft.github.io/vscode-codicons/dist/codicon.html) are supported.
> *Note* that theme icons can also be rendered inside labels and descriptions. Places that support theme icons spell this out and they use the `$(<name>)`-syntax, for instance `quickPick.label = "Hello World $(globe)"`.

akosyakov commented:
Actually please try with the built-in git extension to run `Git checkout` command. It should create a quick pick with plus icon: https://github.com/microsoft/vscode/blob/89906c073a8fc8e60f344c5623f5c73adcf039f1/extensions/git/src/commands.ts#L1585-L1593

RomanNikitenko commented:
@akosyakov 
> Are you sure that you are looking at VS Code with microsoft/vscode@51e5259? Maybe it is already in VS Code insider.

I built VS Code from master branch an hour ago.

> Actually please try with the built-in git extension to run Git checkout command. 
thanks!

akosyakov commented:
> I built VS Code from master branch an hour ago

ok, I think the git status bar is using the custom quick pick, not the command quick palette.

RomanNikitenko commented:
@akosyakov 
I turned off `@theia/git` [here](https://github.com/eclipse-theia/theia/blob/715cda98a8561caf5b3105d63533489dc2ee4ee5/examples/browser/package.json#L27) and added  `vscode-git-1.3.0.1.vsix` into the `plugins` folder.

![git_checkout](https://user-images.githubusercontent.com/5676062/77650554-5586ab80-6f74-11ea-87ed-ac45af343328.gif)

So I think all extensions with `codicons` will be affected.

akosyakov commented:
> So I think all extensions with codicons will be affected.

@RomanNikitenko let's file an issue and have a look at it separately?

RomanNikitenko commented:
@akosyakov 
> @RomanNikitenko let's file an issue and have a look at it separately?

sure!
We are getting `codicons` support within current Monaco upgrade.
A `codicon` is rendered correctly for the most use cases.

To face with the bug a few conditions should be met:
- quick pick
- filtering
- `codicon` variable contains filtering word 

So user never faces with the bug if he/she filters `branch` or `create` and `codicon` is `$(plus)` .

Just tested to confirm that the `codicons` rendering bug has influence on VS Code extensions and we have the reason to investigate how we can fix it in further.
Created the separate issue for that: https://github.com/eclipse-theia/theia/issues/7429

akosyakov commented:
@RomanNikitenko release is done, you can rebase and merge

@azatsarynnyy @RomanNikitenko Can we chat somehow about what you think about finding ways to do https://github.com/eclipse-theia/theia/issues/7100 and https://github.com/eclipse-theia/theia/issues/7261? Instead of upgrading as we do now?

RomanNikitenko commented:
@akosyakov 
> release is done, you can rebase, clean up the history and merge

I have two questions :-)
- should I replace a function [here](https://github.com/eclipse-theia/theia/pull/7149/files#diff-84ba2a793b61b5929c878c84eb4a68f4R234) by method?
- I tried to reorganize git history according to resolved problems. 
But I guess it still too much.
You can see it in the separate branch: https://github.com/eclipse-theia/theia/compare/testRebase

akosyakov commented:
> should I replace the function toCssClassName by method?

it would be nice

> I tried to reorganize git history according to resolved problems.

it looks great

azatsarynnyy commented:
> @azatsarynnyy @RomanNikitenko Can we chat somehow about what you think about finding ways to do #7100 and #7261? Instead of upgrading as we do now?

@akosyakov sure, let's plan it on Monday

akosyakov commented:
@azatsarynnyy Should we meet in our dev meeting room? What time will work for you? at 15 CET?

akosyakov commented:
@RomanNikitenko should anything else be done to merge it?

azatsarynnyy commented:
> @azatsarynnyy Should we meet in our dev meeting room? What time will work for you? at 15 CET?

@akosyakov yes, we can use the same BJN link as for dev call. 15 CET is good for us. Thanks!

RomanNikitenko commented:
@akosyakov 
sorry for the delay - some things on our side...

Thank you very much for your support!
I think this kind of work would be impossible without your assistance.
I really appreciate your help!
